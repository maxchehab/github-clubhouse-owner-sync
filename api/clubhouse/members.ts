import { NextApiResponse, NextApiRequest } from 'next';
import withSession from '../../common/util/with-session.util';

import { bulkUpsertClubhouseMembers } from '../../services/fauna/services/clubhouse-members.service';
import { getClubhouseMembers } from '../../services/clubhouse/clubhouse.service';
import ClubhouseMember from '../../common/interfaces/clubhouse-member.interface';
import lambda from '../../common/util/lambda.util';

export default lambda(
  withSession(async (_req: NextApiRequest, res: NextApiResponse) => {
    const members: ClubhouseMember[] = (
      await getClubhouseMembers(process.env.CLUBHOUSE_SECRET as string)
    ).map(({ profile: { name }, id }) => ({
      id,
      name,
    }));

    await bulkUpsertClubhouseMembers(members);

    return res.json(members);
  }),
);
