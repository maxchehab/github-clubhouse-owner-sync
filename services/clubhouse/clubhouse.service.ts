import Axios from 'axios';
import ClubhouseMember from '../../common/interfaces/clubhouse-member.interface';

interface ClubHouseMember {
  id: string;
  profile: {
    name: string;
  };
}

export async function getClubhouseMembers(
  token: string,
): Promise<ClubHouseMember[]> {
  const { data } = await Axios.get<ClubHouseMember[]>(
    'https://api.clubhouse.io/api/v3/members',
    {
      params: { token },
    },
  );

  return data;
}

export async function updateClubhouseStoryOwner(
  storyID: string,
  owner: ClubhouseMember,
  token: string,
) {
  const { data } = await Axios.put(
    `https://api.clubhouse.io/api/v3/stories/${storyID}`,
    {
      owner_ids: [owner.id],
    },
    {
      params: { token },
    },
  );

  return data;
}
