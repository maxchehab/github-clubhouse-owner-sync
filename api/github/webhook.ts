import withHMAC, { SigParser } from '../../common/util/with-hmac.util';

import { findGitHubMemberByID } from '../../services/fauna/services/github-members.service';
import { updateClubhouseStoryOwner } from '../../services/clubhouse/clubhouse.service';
import getClubhouseStoryID from '../../common/util/get-clubhouse-story-id.util';
import GitHubEvent from '../../common/interfaces/github-event.interface';
import lambda from '../../common/util/lambda.util';

const SECRET = process.env.GITHUB_ORG_HOOK_SECRET as string;

const parser: SigParser = (req, err) => {
  const sig = req.headers['x-hub-signature'] as string;

  if (!sig) {
    err();
  }

  return sig.substring('sha1='.length);
};

export default lambda(
  withHMAC(SECRET, parser, async (req, res) => {
    const event = req.body as GitHubEvent;
    const storyID = getClubhouseStoryID(event.ref);

    if (!storyID) {
      return res.status(200).json({ message: 'Invalid story ID' });
    }

    const member = await findGitHubMemberByID(event.sender.login);

    if (!member) {
      return res.status(200).json({ message: 'Invalid GitHub member' });
    }

    if (!member.clubhouseMember) {
      return res
        .status(200)
        .json({ message: 'No Clubhouse Account linked to GitHub member' });
    }

    await updateClubhouseStoryOwner(
      storyID,
      member.clubhouseMember,
      process.env.CLUBHOUSE_SECRET as string,
    );

    return res.status(200).json({ message: 'Updated story owner' });
  }),
);
