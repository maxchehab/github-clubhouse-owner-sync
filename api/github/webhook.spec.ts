import axios from 'axios';
import http from 'http';
import listen from 'test-listen';

import * as ClubhouseService from '../../services/clubhouse/clubhouse.service';
import * as GitHubMembersService from '../../services/fauna/services/github-members.service';
import closeServer from '../../test/close-server';
import ClubhouseMember from '../../common/interfaces/clubhouse-member.interface';
import createTestLambda from '../../test/create-test-lambda';
import generateHMAC from '../../common/util/generate-hmac.util';
import INVALID_BRANCH_EVENT from '../../test/fixtures/github/invalid-clubhouse-branch.json';
import VALID_BRANCH_EVENT from '../../test/fixtures/github/valid-clubhouse-branch.json';
import webhook from './webhook';

describe('api/github/webhook', () => {
  let server: http.Server;
  let url: string;

  beforeAll(async () => {
    server = createTestLambda(webhook);
    url = await listen(server);
  });

  afterAll(async () => {
    await closeServer(server);
  });

  const SECRET = process.env.GITHUB_ORG_HOOK_SECRET as string;

  describe('with POST api/github/webhook', () => {
    describe('with invalid story ID', () => {
      it('returns invalid story id message', async () => {
        const hmac = generateHMAC(JSON.stringify(INVALID_BRANCH_EVENT), SECRET);

        const { status, data } = await axios.post(url, INVALID_BRANCH_EVENT, {
          headers: {
            'X-Hub-Signature': `sha1=${hmac}`,
          },
        });

        expect(status).toEqual(200);
        expect(data).toMatchSnapshot();
      });
    });

    describe('with invalid GitHub member', () => {
      beforeEach(() => {
        jest
          .spyOn(GitHubMembersService, 'findGitHubMemberByID')
          .mockImplementationOnce(async () => null);
      });

      it('returns invalid GitHub member message', async () => {
        const hmac = generateHMAC(JSON.stringify(VALID_BRANCH_EVENT), SECRET);

        const { status, data } = await axios.post(url, VALID_BRANCH_EVENT, {
          headers: {
            'X-Hub-Signature': `sha1=${hmac}`,
          },
        });

        expect(status).toEqual(200);
        expect(data).toMatchSnapshot();
      });
    });

    describe('with unlinked Clubhouse account', () => {
      beforeEach(() => {
        jest
          .spyOn(GitHubMembersService, 'findGitHubMemberByID')
          .mockImplementationOnce(async () => ({
            id: 'johnsmith',
            avatar: 'https://example.com/avatar.png',
          }));
      });

      it('returns unlinked Clubhouse account message', async () => {
        const hmac = generateHMAC(JSON.stringify(VALID_BRANCH_EVENT), SECRET);

        const { status, data } = await axios.post(url, VALID_BRANCH_EVENT, {
          headers: {
            'X-Hub-Signature': `sha1=${hmac}`,
          },
        });

        expect(status).toEqual(200);
        expect(data).toMatchSnapshot();
      });
    });

    describe('with valid story and GitHub member', () => {
      let updateClubhouseStoryOwner: jest.MockInstance<
        Promise<any>,
        [string, ClubhouseMember, string]
      >;

      beforeEach(() => {
        jest
          .spyOn(GitHubMembersService, 'findGitHubMemberByID')
          .mockImplementationOnce(async () => ({
            id: 'johnsmith',
            avatar: 'https://example.com/avatar.png',
            clubhouseMember: {
              id: '123',
              name: 'John Smith',
            },
          }));

        updateClubhouseStoryOwner = jest
          .spyOn(ClubhouseService, 'updateClubhouseStoryOwner')
          .mockImplementationOnce(async () => null);
      });

      it(`updates story's owner`, async () => {
        const hmac = generateHMAC(JSON.stringify(VALID_BRANCH_EVENT), SECRET);

        const { status, data } = await axios.post(url, VALID_BRANCH_EVENT, {
          headers: {
            'X-Hub-Signature': `sha1=${hmac}`,
          },
        });

        expect(updateClubhouseStoryOwner).toBeCalledTimes(1);
        expect(updateClubhouseStoryOwner.mock.calls[0]).toMatchSnapshot();

        expect(status).toEqual(200);
        expect(data).toMatchSnapshot();
      });
    });
  });
});
