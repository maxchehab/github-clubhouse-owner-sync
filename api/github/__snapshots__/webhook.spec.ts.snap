// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/github/webhook with POST api/github/webhook with invalid GitHub member returns invalid GitHub member message 1`] = `
Object {
  "message": "Invalid GitHub member",
}
`;

exports[`api/github/webhook with POST api/github/webhook with invalid story ID returns invalid story id message 1`] = `
Object {
  "message": "Invalid story ID",
}
`;

exports[`api/github/webhook with POST api/github/webhook with unlinked Clubhouse account returns unlinked Clubhouse account message 1`] = `
Object {
  "message": "No Clubhouse Account linked to GitHub member",
}
`;

exports[`api/github/webhook with POST api/github/webhook with valid story and GitHub member updates story's owner 1`] = `
Array [
  "1161",
  Object {
    "id": "123",
    "name": "John Smith",
  },
  "test-clubhouse-secret",
]
`;

exports[`api/github/webhook with POST api/github/webhook with valid story and GitHub member updates story's owner 2`] = `
Object {
  "message": "Updated story owner",
}
`;
